### DISCUSSION TITLE
Do we still need cuda programming from the perspective of  jax/flax?

### DISCUSSION THREAD

**Discussion by wztdream** [google/flax#1752] [General]
Hi,
This is a general question. 

Sometimes we need some customized operation which is not provided by framework lib, such as [deformable convolution]( https://github.com/google/flax/discussions/1052).  In pytorch or tensorflow this can be done by [cuda programing](https://github.com/fundamentalvision/Deformable-DETR/blob/11169a60c33333af00a4849f1808023eba96a931/models/ops/src/cuda/ms_deform_im2col_cuda.cuh).

Considering jax/flax themself have the mechanism for parallel acceleration (XLA), so arise the quesiton: does it means we do not need cuda programming in jax/flax? How is the performance compered with cuda programming theoretically?

**Comment by avital** [comment]
I'm not sure if the repository is up-to-date but this might help: https://github.com/dfm/extending-jax

**Reply by avital** [reply]
(But in general, some things in JAX don't require custom kernels because XLA knows how to optimize many operations out-of-the-box.)

**Reply by wztdream** [reply]
Thank you, I take it that for most case we do not need cuda programming XLA from jax can handle it properly, but for some edge cases we may need.

**Reply by wztdream** [reply]
I hope one day we can focus only on logic instead of  worrying  about performance too much.
